source/common/unified_memory_allocation/src/memory_pool.c

```diff
@@ -6,43 +6,101 @@
  *
  */
 
+#include "memory_tracker_internal.h"
 #include <uma/memory_pool.h>
+#include <uma/memory_pool_ops.h>
 
 #include <assert.h>
 #include <stdlib.h>
 
 struct uma_memory_pool_t {
-    struct uma_memory_pool_ops_t ops;
     void *pool_priv;
+
+    struct uma_memory_pool_ops_t ops;
+
+    struct uma_memory_provider_desc_t *providers;
+    size_t numProviders;
 };
 
-enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, void *params,
-                                uma_memory_pool_handle_t *hPool) {
+enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, struct uma_memory_provider_desc_t *providers,
+                                size_t numProviders, void *params, uma_memory_pool_handle_t *hPool) {
+    if (providers && !numProviders) {
+        return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!providers && numProviders) {
+        return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    for (size_t i = 0; i < numProviders; i++) {
+        // UMA_MEMORY_PROVIDER_TYPE_TRACKER can only be created by the UMA library
+        if (providers[i].providerType == UMA_MEMORY_PROVIDER_TYPE_TRACKER) {
+            return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    enum uma_result_t ret = UMA_RESULT_SUCCESS;
     uma_memory_pool_handle_t pool = malloc(sizeof(struct uma_memory_pool_t));
     if (!pool) {
         return UMA_RESULT_ERROR_OUT_OF_HOST_MEMORY;
     }
 
     assert(ops->version == UMA_VERSION_CURRENT);
 
-    pool->ops = *ops;
+    pool->providers = calloc(numProviders ? numProviders : 1, sizeof(struct uma_memory_provider_desc_t));
+    if (!pool->providers) {
+        ret = UMA_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_providers_alloc;
+    }
 
-    void *pool_priv;
-    enum uma_result_t ret = ops->initialize(params, &pool_priv);
-    if (ret != UMA_RESULT_SUCCESS) {
-        free(pool);
-        return ret;
+    size_t providerInd = 0;
+    if (!numProviders) {
+        pool->numProviders = 1;
+        pool->providers[0].providerType = UMA_MEMORY_PROVIDER_TYPE_TRACKER;
+        ret = umaTrackingMemoryProviderCreate(NULL, pool, &pool->providers[0].hProvider);
```

Comment:
Maybe for consistency, we should always require at least one memory provider for data?
What is the use case when the user does not provide a memory provider?

Reply:
That would certainly simplify things.

The use case I had in mind is a malloc-based provider (you cannot really implement things like realloc with if you on memory provider level. More generically, the usecase is "any memory pool that forwards calls to some exiting allocator interface that support malloc/realloc/calloc-like API".